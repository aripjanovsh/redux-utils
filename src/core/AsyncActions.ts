import { Action, AnyAction } from "redux";

export interface PerformAction<M = any> extends Action {
  readonly meta: M;
}

export interface FulfillAction<P, M = any> extends PerformAction<M> {
  readonly payload: P;
}

export interface RejectAction<M = any> extends PerformAction<M> {
  readonly error: true;
  readonly payload: Error;
}

export function isPerformAction(action: AnyAction): action is PerformAction {
  return Boolean(
    action &&
      action.meta !== undefined &&
      action.error === undefined &&
      action.payload === undefined,
  );
}

export function createPerformAction<M>(
  type: string,
  meta: M,
): PerformAction<M> {
  return { meta, type };
}

export function isFulfillAction(
  action: AnyAction,
): action is FulfillAction<any> {
  return Boolean(
    action &&
      action.meta !== undefined &&
      action.error === undefined &&
      action.payload !== undefined,
  );
}

export function createFulfillAction<P, M = any>(
  type: string,
  payload: P,
  meta: M,
): FulfillAction<P, M> {
  return { meta, type, payload };
}

export function isRejectAction(action: AnyAction): action is RejectAction {
  return Boolean(
    action &&
      action.error === true &&
      action.meta !== undefined &&
      action.payload !== undefined,
  );
}

export function createRejectAction<M = any>(
  type: string,
  error: Error,
  meta: M,
): RejectAction<M> {
  return { meta, type, error: true, payload: error };
}
