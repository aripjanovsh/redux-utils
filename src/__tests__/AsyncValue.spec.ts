import { Action, AnyAction } from "redux";

import { FulfillAction, PerformAction, RejectAction } from "../AsyncActions";
import {
  createAsyncValueDictReducer,
  createAsyncValueReducer,
  getInitialAsyncValue,
} from "../AsyncValue";

describe("AsyncRequest", () => {
  const perform = "perform";
  const fulfill = "fulfill";
  const reject = "reject";
  const reset = "reset";

  describe("getInitialAsyncRequest", () => {
    test("initial value", () => {
      expect(getInitialAsyncValue()).toMatchSnapshot();
    });
  });

  describe("createAsyncRequestReducer", () => {
    test("creation", () => {
      const reducer = createAsyncValueReducer<number>(
        perform,
        fulfill,
        reject,
        reset,
      );

      expect(typeof reducer).toBe("function");
    });

    test("perform action", () => {
      const reducer = createAsyncValueReducer<number>(
        perform,
        fulfill,
        reject,
        reset,
      );

      const error = new Error("Async Error");
      const action: PerformAction = { meta: {}, type: perform };

      expect(reducer({ fetching: false }, action)).toEqual({
        fetching: true,
      });

      expect(reducer({ value: 10, fetching: false }, action)).toEqual({
        value: 10,
        fetching: true,
      });

      expect(reducer({ error, fetching: false }, action)).toEqual({
        error,
        fetching: true,
      });

      expect(reducer({ error, value: 10, fetching: false }, action)).toEqual({
        error,
        value: 10,
        fetching: true,
      });
    });

    test("fulfill action", () => {
      const reducer = createAsyncValueReducer<number>(
        perform,
        fulfill,
        reject,
        reset,
      );

      const error = new Error("Async Error");
      const action: FulfillAction<number> = {
        meta: {},
        payload: 30,
        type: fulfill,
      };

      expect(reducer({ fetching: true }, action)).toEqual({
        value: 30,
        fetching: false,
      });

      expect(reducer({ value: 10, fetching: true }, action)).toEqual({
        value: 30,
        fetching: false,
      });

      expect(reducer({ error, fetching: true }, action)).toEqual({
        value: 30,
        fetching: false,
      });

      expect(reducer({ error, value: 10, fetching: true }, action)).toEqual({
        value: 30,
        fetching: false,
      });
    });

    test("reject action", () => {
      const reducer = createAsyncValueReducer<number>(
        perform,
        fulfill,
        reject,
        reset,
      );

      const stateError = new Error("AsyncError: State.");
      const actionError = new Error("AsyncError: Action.");
      const action: RejectAction = {
        meta: {},
        error: true,
        type: reject,
        payload: actionError,
      };

      expect(reducer({ fetching: true }, action)).toEqual({
        fetching: false,
        error: actionError,
      });

      expect(reducer({ value: 10, fetching: true }, action)).toEqual({
        value: 10,
        fetching: false,
        error: actionError,
      });

      expect(reducer({ fetching: true, error: stateError }, action)).toEqual({
        fetching: false,
        error: actionError,
      });

      expect(
        reducer({ value: 10, fetching: true, error: stateError }, action),
      ).toEqual({
        value: 10,
        fetching: false,
        error: actionError,
      });
    });

    test("reset action", () => {
      const reducer = createAsyncValueReducer<number>(
        perform,
        fulfill,
        reject,
        reset,
      );

      const error = new Error("AsyncError: State.");
      const action: Action = { type: reset };

      expect(reducer({ fetching: true }, action)).toEqual({
        fetching: false,
      });

      expect(reducer({ value: 10, fetching: true }, action)).toEqual({
        fetching: false,
      });

      expect(reducer({ error, fetching: true }, action)).toEqual({
        fetching: false,
      });

      expect(reducer({ error, value: 10, fetching: true }, action)).toEqual({
        fetching: false,
      });
    });

    test("unknown action", () => {
      const reducer = createAsyncValueReducer<number>(
        perform,
        fulfill,
        reject,
        reset,
      );

      const error = new Error("AsyncError: State.");
      const action: Action = { type: "FOO" };

      expect(
        reducer(
          {
            fetching: true,
          },
          action,
        ),
      ).toEqual({ fetching: true });

      expect(reducer({ value: 10, fetching: true }, action)).toEqual({
        value: 10,
        fetching: true,
      });

      expect(reducer({ error, fetching: true }, action)).toEqual({
        error,
        fetching: true,
      });

      expect(reducer({ error, value: 10, fetching: true }, action)).toEqual({
        error,
        value: 10,
        fetching: true,
      });
    });
  });

  describe("createAsyncValueDictReducer", () => {
    interface ActionMeta {
      readonly id: number;
    }

    const hashResolver = (action: AnyAction) =>
      (action as PerformAction<ActionMeta>).meta.id;

    test("creation", () => {
      const reducer = createAsyncValueDictReducer<number>(
        hashResolver,
        perform,
        fulfill,
        reject,
        reset,
      );

      expect(typeof reducer).toBe("function");
    });

    test("unknown action", () => {
      const reducer = createAsyncValueDictReducer<number>(
        hashResolver,
        perform,
        fulfill,
        reject,
        reset,
      );

      const action: Action = { type: "FOO" };

      expect(reducer({}, action)).toEqual({});
    });

    test("flow", () => {
      const reducer = createAsyncValueDictReducer<number>(
        hashResolver,
        perform,
        fulfill,
        reject,
        reset,
      );

      let state = {};

      state = reducer(state, {
        type: perform,
        meta: { id: 1 },
      } as PerformAction<ActionMeta>);

      expect(state).toEqual({ 1: { fetching: true } });

      state = reducer(state, {
        payload: 10,
        type: fulfill,
        meta: { id: 1 },
      } as FulfillAction<number, ActionMeta>);

      expect(state).toEqual({ 1: { fetching: false, value: 10 } });

      state = reducer(state, {
        payload: 20,
        type: perform,
        meta: { id: 2 },
      } as FulfillAction<number, ActionMeta>);

      expect(state).toEqual({
        1: { fetching: false, value: 10 },
        2: { fetching: true },
      });

      const error = new Error("AsyncError");

      state = reducer(state, {
        payload: error,
        type: reject,
        meta: { id: 2 },
        error: true,
      } as RejectAction<ActionMeta>);

      expect(state).toEqual({
        1: { fetching: false, value: 10 },
        2: { fetching: false, error },
      });

      state = reducer(state, {
        type: perform,
        meta: { id: 2 },
      } as PerformAction<ActionMeta>);

      expect(state).toEqual({
        1: { fetching: false, value: 10 },
        2: { fetching: true, error },
      });

      state = reducer(state, {
        payload: 20,
        type: perform,
        meta: { id: 2 },
      } as FulfillAction<number, ActionMeta>);

      expect(state).toEqual({
        1: { fetching: false, value: 10 },
        2: { fetching: true, error },
      });
    });
  });
});
