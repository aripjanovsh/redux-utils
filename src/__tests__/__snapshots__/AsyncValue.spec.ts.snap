// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncRequest getInitialAsyncRequest initial value 1`] = `
Object {
  "fetching": false,
}
`;
